### PR TITLE
Flavor and build type support added

### PR DESCRIPTION
**Pull request by piwonski**

### CODE REVIEW

File: CockpitPlugin/src/main/java/com/polidea/cockpitplugin/CockpitTask.kt [hunk 1]
```diff
@@ -2,62 +2,126 @@ package com.polidea.cockpitplugin
...
+    private fun parseValues(values: Map<String, Any>): ArrayList<Param<*>> {
         val paramList = ArrayList<Param<*>>()
 
         values.map {
-            when (it.value) {
-                is String -> paramList.add(StringParam(it.key, it.value as String))
-                is Int -> paramList.add(IntegerParam(it.key, it.value as Int))
-                is Double -> paramList.add(DoubleParam(it.key, it.value as Double))
-                is Boolean -> paramList.add(BooleanParam(it.key, it.value as Boolean))
+            val value = it.value
```
Comment (martawoldanska) [inline CockpitPlugin/src/main/java/com/polidea/cockpitplugin/CockpitTask.kt]: It seems a bit inconsistent that it.value is extracted, but it.key is not

File: sample/src/debug/java/com/polidea/cockpit/sample/util/Extensions.kt [hunk 2]
```diff
@@ -9,6 +10,7 @@ internal fun MainActivity.initViews() {
     findViewById<View>(R.id.edit_values_button).setOnClickListener {
```
Comment (MarcinKornacki) [inline sample/src/debug/java/com/polidea/cockpit/sample/util/Extensions.kt]: This is not a proper extension. What about creating BaseMainActivity in both variants and implement initView() method in their bodies? findViewById call won't be necessary.

File: CockpitPlugin/src/main/java/com/polidea/cockpitplugin/util/Util.kt [hunk 3]
```diff
@@ -0,0 +1,42 @@
+package com.polidea.cockpitplugin.util
```
Comment (MarcinKornacki) [inline CockpitPlugin/src/main/java/com/polidea/cockpitplugin/util/Util.kt]: reformat file

File: CockpitPlugin/src/test/java/com/polidea/cockpitplugin/util/UtilTest.kt [hunk 4]
```diff
@@ -0,0 +1,261 @@
+package com.polidea.cockpitplugin.util
```
Comment (MarcinKornacki) [inline CockpitPlugin/src/test/java/com/polidea/cockpitplugin/util/UtilTest.kt]: reformat file

File: sample/src/debug/res/layout/include_edit_values_container.xml [hunk 5]
```diff
@@ -1,12 +1,29 @@
 <?xml version="1.0" encoding="utf-8"?>
```
Comment (MarcinKornacki) [inline sample/src/debug/res/layout/include_edit_values_container.xml]: reformat file

File: sample/src/main/res/layout/activity_main.xml [hunk 6]
```diff
@@ -10,29 +10,42 @@
         android:id="@+id/cockpit_textview"
```
Comment (MarcinKornacki) [inline sample/src/main/res/layout/activity_main.xml]: reformat file

File: CockpitPlugin/src/main/java/com/polidea/cockpitplugin/input/InputFilesProvider.kt [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        if (dimensions.isNotEmpty()) {
+            buildTypes.find { variantName.contains(it, ignoreCase = true) }?.let {
+                ordered.add(it.capitalize())
+            } // debug
+        }
+
+        ordered.add(variantName.capitalize()) // stagingDemoDebug
+
+        return ordered.map { fileFactory.file(cockpitDirectoryPath + "cockpit" + it + ".yml") }
+                .filter { fileFactory.isFileExists(it) }
```
Comment (MarcinKornacki) [inline CockpitPlugin/src/main/java/com/polidea/cockpitplugin/input/InputFilesProvider.kt]: Just use `it.exists()`

File: CockpitPlugin/src/test/java/com/polidea/cockpitplugin/input/InputFileProviderTest.kt [hunk 8]
```diff
@@ -0,0 +1,527 @@
...
+    }
+
+    private fun createInputFilesProvider(fileExistsClosure: (File) -> Boolean): InputFilesProvider {
+        val fileFactory = createFileFactory(fileExistsClosure)
+        return InputFilesProvider("directory/", fileFactory)
+    }
+
+    companion object {
+
+        val noDimensions = listOf<String>()
```
Comment (MarcinKornacki) [inline CockpitPlugin/src/test/java/com/polidea/cockpitplugin/input/InputFileProviderTest.kt]: `emptyList<String>()`

File: CockpitPlugin/src/test/java/com/polidea/cockpitplugin/input/InputFileProviderTest.kt [hunk 9]
```diff
@@ -0,0 +1,527 @@
...
+        val dimensionsApi = listOf("api")
+        val dimensionsApiMode = listOf("api", "mode")
+        val dimensionsApiModeReversed = dimensionsApiMode.reversed()
+
+        val noFlavors = listOf<Flavor>()
```
Comment (MarcinKornacki) [inline CockpitPlugin/src/test/java/com/polidea/cockpitplugin/input/InputFileProviderTest.kt]: `emptyList< Flavor >()`

File: CockpitPlugin/src/test/java/com/polidea/cockpitplugin/util/UtilTest.kt [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+                Pair("p2", false),
+                Pair("p3", 0)
+        )
+        assertEquals(expectedMap, mergedMap)
+    }
+
+    @Test
+    fun createUtilEmptySetTest() {
+
+        val mainMap = mapOf<String, Any>().toMutableMap()
```
Comment (MarcinKornacki) [inline CockpitPlugin/src/test/java/com/polidea/cockpitplugin/util/UtilTest.kt]: `emptyMap<String, Any>().toMutableMap()`

File: CockpitPlugin/src/test/java/com/polidea/cockpitplugin/util/UtilTest.kt [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+
+        val mergedMap = Util.deepMerge(mainMap)
+        val expectedMap = mapOf<String, Any>()
```
Comment (MarcinKornacki) [inline CockpitPlugin/src/test/java/com/polidea/cockpitplugin/util/UtilTest.kt]: `emptyMap<String, Any>()`

File: CockpitPlugin/src/main/java/com/polidea/cockpitplugin/YamlReaderAndWriter.kt [hunk 12]
```diff
@@ -0,0 +1,28 @@
+package com.polidea.cockpitplugin
+
+import org.yaml.snakeyaml.LoaderOptions
+import org.yaml.snakeyaml.Yaml
+import java.io.File
+import java.io.FileWriter
+
+class YamlReaderAndWriter {
```
Comment (martawoldanska) [inline CockpitPlugin/src/main/java/com/polidea/cockpitplugin/YamlReaderAndWriter.kt]: <3

File: CockpitPlugin/src/main/java/com/polidea/cockpitplugin/input/InputFilesProvider.kt [hunk 13]
```diff
@@ -6,39 +6,140 @@ import java.io.File
...
+        //                dimension "api"
+        //                applicationIdSuffix ".prod"
+        //            }
+        //            staging {
+        //                dimension "api"
+        //                applicationIdSuffix ".staging"
+        //            }
+        //        }
+        //
+        // Let's assume also, we are currently building a 'stagingDemoDebug' variant.
```
Comment (martawoldanska) [inline CockpitPlugin/src/main/java/com/polidea/cockpitplugin/input/InputFilesProvider.kt]: Let's rephrase it to "Let's assume as well that we are currently (...)"

File: CockpitPlugin/src/main/java/com/polidea/cockpitplugin/input/InputFilesProvider.kt [hunk 14]
```diff
@@ -6,39 +6,140 @@ import java.io.File
...
+        // Let's take a look, what parameters will take our function:
```
Comment (martawoldanska) [inline CockpitPlugin/src/main/java/com/polidea/cockpitplugin/input/InputFilesProvider.kt]: Let's rephrase to "Let's take a look what parameters our function will take"

File: CockpitPlugin/src/main/java/com/polidea/cockpitplugin/input/InputFilesProvider.kt [hunk 15]
```diff
@@ -6,39 +6,140 @@ import java.io.File
...
+        // dimensions == ["api", "mode"]
+        // flavors == [Flavor("demo", "mode"), Flavor("full", "mode"), Flavor("prod", "api"), Flavor("staging", "api")
+        // variantName == "stagingDemoDebug"
+        // buildTypes == ["debug", "release"]
+        //
+        // According to:
+        //
+        // https://developer.android.com/studio/write/add-resources#resource_merging
+        //
+        // resource merging prority is as follows:
```
Comment (martawoldanska) [inline CockpitPlugin/src/main/java/com/polidea/cockpitplugin/input/InputFilesProvider.kt]: Typo: priority

File: CockpitPlugin/src/main/java/com/polidea/cockpitplugin/input/InputFilesProvider.kt [hunk 16]
```diff
@@ -6,39 +6,140 @@ import java.io.File
...
+        // Someone could ask a question: "Why `staging` flavor is first in variant's name `stagingDemoDebug`?
+        // Why the variant name is `stagingDemoDebug` not `demoStagingDebug`?" The name (and flavors' priority) is determined by
+        // order of flavorDimensions in build.gradle file. In our case this is:
+        //
+        //        flavorDimensions "api", "mode"
+        //
+        // `staging` belongs to `api` dimension, so it's the first
+        // `demo` belongs to `mode` dimesion, so it's the second in `stagingDemoDebug` variant's name
+        //
+        // Finally, resource merging priority is following:
```
Comment (martawoldanska) [inline CockpitPlugin/src/main/java/com/polidea/cockpitplugin/input/InputFilesProvider.kt]: Let's change "following" to "as follows"

File: CockpitPlugin/src/main/java/com/polidea/cockpitplugin/input/InputFilesProvider.kt [hunk 17]
```diff
@@ -6,39 +6,140 @@ import java.io.File
...
+        val variantsOrderedFromLowestPriority = ArrayList<String>()
+
+        // `main` has the lowest priority, so we add it as a first
+        // Empty string corresponds with `main` variant
+        // because cockpit.yml doesn't contains "main" string in its name
+        variantsOrderedFromLowestPriority.add("")
+
+        // Next, the lowest priority has `demo` and `staging` flavors.
+        //
+        // Declared dimension's order is following:
```
Comment (martawoldanska) [inline CockpitPlugin/src/main/java/com/polidea/cockpitplugin/input/InputFilesProvider.kt]: Let's change "following" to "as follows"

File: CockpitPlugin/src/main/java/com/polidea/cockpitplugin/input/InputFilesProvider.kt [hunk 18]
```diff
@@ -6,39 +6,140 @@ import java.io.File
...
+        //
+        //        flavorDimensions "api", "mode"
+        //
+        // so we have to reverse dimensions list, to add `demo` (from `mode` dimension) first
+        // and `staging` (from `api` dimension) after `demo`
+        //
+        // Only `staging` and `demo` flavors should be considered, because only those two flavors appear in
+        // variant `stagingDemoDebug`. So we find only those variants which appears in variant's name.
```
Comment (martawoldanska) [inline CockpitPlugin/src/main/java/com/polidea/cockpitplugin/input/InputFilesProvider.kt]: "So we find only those variants which appears in variant's name." – I guess you meant "So we find only those flavors which appear in variant's name". :)

File: CockpitPlugin/src/main/java/com/polidea/cockpitplugin/input/InputFilesProvider.kt [hunk 19]
```diff
@@ -6,39 +6,140 @@ import java.io.File
...
 
-        ordered.addAll(singleFlavors)
+        variantsOrderedFromLowestPriority.addAll(singleFlavors)
 
+        // Composite flavor exists only when application has more than 1 flavor dimension
         if (dimensions.size > 1) {
+
+            // Next, the lowest priority has `stagingDemo` flavor
+            //
+            // singleFlavors is equals ["Demo", "Staging"], so we have to reverse again the list
```
Comment (martawoldanska) [inline CockpitPlugin/src/main/java/com/polidea/cockpitplugin/input/InputFilesProvider.kt]: It's either "is equal" or "equals". ;)

Also please rephrase: "(...) so we have to reverse again the list" to "(...) so we have to reverse the list again"

File: CockpitPlugin/src/main/java/com/polidea/cockpitplugin/input/InputFilesProvider.kt [hunk 20]
```diff
@@ -6,39 +6,140 @@ import java.io.File
...
+
+            // Next, the lowest priority has build type.
             buildTypes.find { variantName.contains(it, ignoreCase = true) }?.let {
-                ordered.add(it.capitalize())
+                variantsOrderedFromLowestPriority.add(it.capitalize())
             } // debug
         }
 
-        ordered.add(variantName.capitalize()) // stagingDemoDebug
+        // At the end, we should add variantName as a last item, as variantName has the highest priority in
```
Comment (martawoldanska) [inline CockpitPlugin/src/main/java/com/polidea/cockpitplugin/input/InputFilesProvider.kt]: Please change "a last item" to "the last item"

File: CockpitPlugin/src/main/java/com/polidea/cockpitplugin/input/InputFilesProvider.kt [hunk 21]
```diff
@@ -6,39 +6,140 @@ import java.io.File
...
+        // resource merging order
+        variantsOrderedFromLowestPriority.add(variantName.capitalize()) // stagingDemoDebug
 
-        return ordered.map { fileFactory.file(cockpitDirectoryPath + "cockpit" + it + ".yml") }
-                .filter { fileFactory.isFileExists(it) }
+        // Now we converts ["", "Demo", "Staging", "DemoStaging", ...] list to
```
Comment (martawoldanska) [inline CockpitPlugin/src/main/java/com/polidea/cockpitplugin/input/InputFilesProvider.kt]: Typo: convert

File: sample/src/debug/java/com/polidea/cockpit/sample/MainActivity.kt [hunk 22]
```diff
@@ -0,0 +1,13 @@
+package com.polidea.cockpit.sample
+
+import android.view.View
+import com.polidea.cockpit.cockpit.Cockpit
+
+class MainActivity : VariantIndependentBaseMainActivity() {
```
Comment (martawoldanska) [inline sample/src/debug/java/com/polidea/cockpit/sample/MainActivity.kt]: 👏
